fathom/src/surface/elaboration.rs

```diff
@@ -1199,8 +1211,28 @@ impl<'interner, 'arena> Context<'interner, 'arena> {
             }
             (Term::ReportedError(range), _) => core::Term::Prim(range.into(), Prim::ReportedError),
             (_, _) => {
-                let (core_term, synth_type) = self.synth(surface_term);
-                self.convert(surface_term.range(), core_term, &synth_type, &expected_type)
+                let surface_range = surface_term.range();
+                let (mut synth_term, mut synth_type) = self.synth(surface_term);
+
+                while let (
+                    Value::FunType(Plicity::Implicit, name, param_type, body_type),
+                    Value::FunType(Plicity::Explicit, ..),
+                ) = (
+                    self.elim_env().force(&synth_type).as_ref(),
+                    expected_type.as_ref(),
+                ) {
+                    let source = MetaSource::ImplicitArg(surface_range, *name);
+                    let arg_term = self.push_unsolved_term(source, param_type.clone());
+                    let arg_value = self.eval_env().eval(&arg_term);
+                    synth_term = core::Term::FunApp(
+                        surface_range.into(),
+                        Plicity::Implicit,
+                        self.scope.to_scope(synth_term),
+                        self.scope.to_scope(arg_term),
+                    );
+                    synth_type = self.elim_env().apply_closure(body_type, arg_value);
+                }
```

Comment:
I think implicit insertion doesn't need to be done on other eliminators. After all, there's no such thing as implicit pattern matching or implicit record projections. I could be wrong though.

Reply:
I was thinking about what happens if you had something like `foo : {A} -> { x : A }`, then you did `foo.x`?

Reply:
You would get a type error because you are trying to do a record projection on a function. You would have to provide the argument explicitly, eg `(foo @Bool).x`, or use `foo` in a checked context first, eg `let bar : Bool = foo; bar.x`.

Are there any scenarios where inserting implicit arguments before a projection would be helpful? Maybe a parameterized module?

Reply:
Not sure about use cases… I guess it just seems like a natural place to insert implicit args? Because eliminations (excluding applications) can't be applied to functions you can try to insert implicit args until you get to something that looks like what you were expecting, or report an error.